src/blockchain/compact_filters/p2p.rs

```diff
@@ -0,0 +1,646 @@
+            Network::Signet => &[],
+        };
+
+        for seedhost in seedhosts.iter() {
+            if let Ok(lookup) = (*seedhost, port).to_socket_addrs() {
+                for host in lookup {
+                    if host.is_ipv4() {
+                        seeds.push_back(host);
+                    }
+                }
+            }
+        }
+        seeds
+    }
+}
+
+/// Crawler structure that will interface with Discovery and public bitcoin network
+///
+/// Address manager will spawn multiple crawlers in separate threads to discover new addresses.
+struct AddressWorker {
+    discovery: Arc<RwLock<AddressDiscovery>>,
+    sender: Sender<AddressMessage>,
+    network: Network,
+}
+
+impl AddressWorker {
+    fn new(
+        discovery: Arc<RwLock<AddressDiscovery>>,
+        sender: Sender<AddressMessage>,
+        network: Network,
+    ) -> AddressWorker {
+        AddressWorker {
+            discovery,
+            sender,
+            network,
+        }
+    }
+
+    fn try_receive_addr(&mut self, peer: &Peer) {
+        if let Some(NetworkMessage::Addr(new_addresses)) =
+            peer.recv("addr", Some(Duration::from_secs(1))).unwrap()
+        {
+            self.consume_addr(new_addresses);
+        }
+    }
+
+    fn consume_addr(&mut self, addrs: Vec<(u32, Address)>) {
+        let mut discovery_lock = self.discovery.write().unwrap();
+        let mut addresses = Vec::new();
+        for network_addrs in addrs {
+            if let Ok(socket_addrs) = network_addrs.1.socket_addr() {
+                addresses.push(socket_addrs);
+            }
+        }
+        discovery_lock.add_pendings(addresses);
+    }
+
+    fn work(&mut self) {
+        loop {
+            let next_address = {
+                let mut address_discovery = self.discovery.write().unwrap();
+                address_discovery.get_next()
+            };
+
+            match next_address {
+                Some(address) => {
+                    let potential_peer = Peer::connect_with_timeout(
+                        address,
+                        Duration::from_secs(1),
+                        Arc::new(Mempool::default()),
+                        self.network,
+                    );
+
+                    if let Ok(peer) = potential_peer {
+                        peer.send(NetworkMessage::GetAddr).unwrap();
+                        self.try_receive_addr(&peer);
+                        self.sender
+                            .send(AddressMessage::NewValid(
+                                address,
+                                peer.get_version().services,
+                            ))
+                            .unwrap();
+                        // TODO: Investigate why close is being called on non existent connections
+                        // currently the errors are ignored
+                        peer.close().unwrap_or(());
+                    }
+                }
+                None => continue,
```

Comment:
Maybe if we can't connect we should "ban" the peer?

I wonder if it makes sense to add a "ban until" timestamp to the cache, so that we can differentiate between a permanent ban (peer misbehaving) or a temporary "i can't connect right now" ban

Reply:
Hmm, probably. Currently what it does is if it can't connect it will simply skip. I am not sure if connection failure should result into ban, as nodes can be offline temporarily, they may connect in next try. 

I think we definitely need to define banning more granularly, and I think best place to do that will be the `peermanager` (which I am sketching now, and it will use address manager inside). I have put a ban score flag into PeerManager, so we can have more granular banning logic like core does. 

And the objective here is the cache will maintain only the final ban list such that once it's banned it will never get into discovery again.

And then in peer manager, we can design different levels of banning, and decide whether to tell the address manager to "cache ban" the address or not.